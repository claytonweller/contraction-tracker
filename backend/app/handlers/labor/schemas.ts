import * as z from 'zod'

const numberArray = z.array(z.number())

export const contractionSchema = z.object({
  startTime: z.string(),
  endTime: z.string().optional(),
  intensity: z.number().optional()
})

export const calculatedLaborSchema = z.object({
  contraction:  z.object({
    averageDuration: z.number(),
    averageIntensity: z.number(),
    durations: numberArray,
    averageDurations: numberArray,
    intensities: numberArray
  }),
  rest: z.object({
    currentDuration: z.number().optional(),
    averageDuration: z.number(),
    durations: numberArray,
  }),
  isGoTime: z.boolean(),
  contractionsAreLongEnough: z.boolean(), 
  restsAreShortEnough: z.boolean(), 
  laborIsLongEnough: z.boolean()
})

export const laborSchema = z.object({
  userId: z.string(),
  startTime: z.string(),
  endTime: z.string().optional(),
  contractions:z.array(contractionSchema),
  calculated: calculatedLaborSchema,
  bishopScore: z.number()
})