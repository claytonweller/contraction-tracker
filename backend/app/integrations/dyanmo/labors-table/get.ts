import { ILabor } from "../../../types/Labor";
import { IBaseDynamoParams } from "../index";

export async function getLabors(
  params: { userId: string },
  client: AWS.DynamoDB.DocumentClient,
  baseParams: IBaseDynamoParams
): Promise<ILabor[]> {
  const { userId } = params;
  const getParams = {
    ...baseParams,
    Key: { userId }
  }
  const response = await client.get(getParams).promise()
  const existingLabors: ILabor[] = response?.Item?.labors || []
  return existingLabors
}