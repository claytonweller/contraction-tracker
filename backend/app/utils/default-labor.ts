import { DateTime } from "luxon";
import { IContraction } from "../types/Contraction";
import { ILabor } from "../types/Labor";

export const defaultLabor = (params:{
  totalMinutesOfLabor?: number,
  contractions?: IContraction[],
}={}): ILabor => {
  const now = DateTime.now()
  const {
    totalMinutesOfLabor = 30,
    contractions = [
      {
        startTime: now.minus({ minute: 25 }).toISO(),
        endTime: now.minus({ minute: 24 }).toISO(),
        intensity: 1
      },
      {
        startTime: now.minus({ minute: 20 }).toISO(),
        endTime: now.minus({ minute: 18 }).toISO(),
        intensity: 2
      }

    ]
  } = params
  
  return {
    userId: "1",
    startTime: now.minus({ minute: totalMinutesOfLabor }).toISO(),
    endTime: undefined,
    contractions,
    calculated: {
      isGoTime: false,
      contraction: {
        averageDuration: 6000,
        averageIntensity: 1,
        durations: [],
        intensities: []
      },
      rest: {
        currentDuration: 6000,
        averageDuration: 6000,
        durations: [],
      }
    },
    bishopScore: 0
  }
}

