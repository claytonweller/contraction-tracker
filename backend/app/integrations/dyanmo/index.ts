import * as AWS from 'aws-sdk'
import { connectToLaborTable } from "./labors-table";
import { connectToUserTable } from "./users-table";

const client = new AWS.DynamoDB.DocumentClient()

export const dynamo = {
  labor: connectToLaborTable(client),
  user: connectToUserTable(client)
}

export interface IBaseDynamoParams {
  TableName: string
}