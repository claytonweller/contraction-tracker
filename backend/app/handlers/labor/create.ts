import { withErrorHandling } from "../../utils/with-error-handling"
import { dynamo } from "../../integrations/dyanmo"
import { calculateLaborValues } from "../../utils/calculate-labor-values";

async function createHandler(event) {
  console.log('CREATE', event)
  const labor = JSON.parse(event.body)
  const laborWithCalculatedValues = calculateLaborValues(labor)
  const createdLabor = await dynamo.labor.create(laborWithCalculatedValues, "123")
  return createdLabor
}

export const handler = withErrorHandling(createHandler)
