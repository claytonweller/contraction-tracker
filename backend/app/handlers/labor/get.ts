import { dynamo } from "../../integrations/dyanmo"
import { withErrorHandling } from "../../utils/with-error-handling"
import {z} from 'zod'

export async function getHandler(event) {
  const {userId} = event.queryStringParameters
  const validUserId = z.string().parse(userId)
  const labors = await dynamo.labor.get(validUserId)
  return labors
}

export const handler = withErrorHandling(getHandler)