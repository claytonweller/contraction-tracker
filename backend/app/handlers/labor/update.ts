import { dynamo } from "../../integrations/dyanmo"
import { calculateLaborValues } from "../../utils/calculate-labor-values"
import { withErrorHandling } from "../../utils/with-error-handling"
import {laborSchema} from '../../types/Labor'

async function updateHandler(event) {
  const labor = JSON.parse(event.body)
  const validLabor = laborSchema.parse(labor)
  const laborWithCalculations = calculateLaborValues(validLabor)
  const updatedLabor = await dynamo.labor.updateActive(laborWithCalculations)
  return updatedLabor
}

export const handler = withErrorHandling(updateHandler)