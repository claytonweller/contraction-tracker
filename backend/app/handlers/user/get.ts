import { dynamo } from "../../integrations/dyanmo"
import { withErrorHandling } from "../../utils/with-error-handling"

export async function getHandler(event) {
  const {userId} = event.queryStringParameters
  const user = await dynamo.user.get(userId)
  return user
}

export const handler = withErrorHandling(getHandler)