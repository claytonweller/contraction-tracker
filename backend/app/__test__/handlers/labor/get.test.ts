import { getHandler as getLaborsHandler } from "../../../handlers/labor/get";
import { dynamo } from "../../../integrations/dyanmo";
import { defaultLabor } from "../../../utils/default-labor";

describe('updateLaborHandler', ()=>{
  const dynamoSpy = jest.spyOn(dynamo.labor, 'get')
  beforeEach(()=>{
    jest.clearAllMocks()
  })
  const validLabor = {...defaultLabor(), endTime: undefined}

  describe('when the body is incorrectly formatted', ()=>{
    const incorrect = {queryStringParameters:{noUserId: 'to be found'}}
    let error 
    beforeEach(async ()=>{
      try {  
        await getLaborsHandler(incorrect)
      } catch (e){
        error = e
      }
    })  
    
    it('should throw an error', ()=>{
      expect(error).toMatchSnapshot()
    })

    it('should not attempt to get labors', ()=>{
      expect(dynamoSpy).not.toHaveBeenCalled()
    })
  })

  describe('when the body is correctly formatted', ()=>{
    const userId = 'is present'
    const correct = {queryStringParameters:{userId}}

    describe('and the request for labors succeeds', ()=>{
      let result
      beforeEach(async()=>{
        dynamoSpy.mockResolvedValue([validLabor])
        result = await getLaborsHandler(correct)
      })
      it('should return a correctly formatted labor', ()=>{
        expect(result).toEqual([validLabor])
      })
      it('should have attempted to persist the labor', ()=>{
        expect(dynamoSpy).toHaveBeenCalledWith(userId)
      })
    })

    describe('and the request fails', ()=>{
      const message = 'KABOOM!'
      let error 
      beforeEach(async ()=>{
        dynamoSpy.mockRejectedValue(message)
        try {  
          await getLaborsHandler(correct)
        } catch (e){
          error = e
        }
      })  
      it('should throw an error', ()=>{
        expect(error).toBe(message)
      })
      it('should have attempted to persist the labor', ()=>{
        expect(dynamoSpy).toHaveBeenCalledWith(userId)
      })
    })
  })
})