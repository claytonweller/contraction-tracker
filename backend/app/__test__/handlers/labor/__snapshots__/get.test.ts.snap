// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`updateLaborHandler when the body is incorrectly formatted should throw an error 1`] = `
[ZodError: [
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "undefined",
    "path": [],
    "message": "Required"
  }
]]
`;
